### PR TITLE
Implement AWS Bedrock provider for AI Gateway

### PR DESCRIPTION
**Pull request by andrew-christianson**
## Related Issues/PRs

N/A

## What changes are proposed in this pull request?

This PR implements an MLflow AI Gateway `BaseProvider` for [AWS Bedrock](https://aws.amazon.com/bedrock/).  This exposes AWS' [Titan](https://aws.amazon.com/bedrock/titan/) model as well as models from other providers offered via AWS Bedrock through AI Gateway. 

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

This PR adds new unit tests and an integration test case exercising the provider-specific pieces.  I also tested full calls to Bedrock in an included, marked-skip test case.

N.B. Bedrock is still in preview, so live tests depend on an AWS account with access to the preview.  Also, support is not yet included in published versions of botocore, so live tests (and deployments) will need to install boto3/botocore versions linked from AWS Bedrock

## Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [x] Instructions

## Release Notes

Support for AWS Bedrock in MLflow AI Gateway.  

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add support for AWS Bedrock in MLflow AI Gateway.  AI Gateway users can use Bedrock to access Amazon's Titan LLM, as well as models from third-party providers.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/providers/bedrock.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+                    payload,
+                    {
+                        "stop": "stop_sequences",
+                        "candidate_count": "num_generations",
+                    },
+                ),
+            }
+            # The range of Cohere's temperature is 0-5, but ours is 0-1, so we scale it.
+            if "temperature" in preped:
+                preped["temperature"] *= 5
```
Comment (dbczumar) [inline mlflow/gateway/providers/bedrock.py:184]: Is there an opportunity for us to reuse the payload translation logic from https://github.com/mlflow/mlflow/blob/61d8622dde16578dd844ecc12668580dd81a311c/mlflow/gateway/providers/cohere.py#L43-L44, etc? If possible, could we extract these out into some common utility functions that we can call in each provider (Bedrock for cohere, cohere, bedrock for anthropic, anthropic, etc)?

File: mlflow/gateway/providers/bedrock.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    def _process_output(self, response: dict):
+        if self._underlying_provider is AWSBedrockModelProvider.ANTHROPIC:
+            return {
+                "candidates": [
+                    {
+                        "text": response.get("completion"),
+                        "metadata": {
+                            "finish_reason": "stop"
+                            if response.get("stop_reason") == "stop_sequence"
```
Comment (dbczumar) [inline mlflow/gateway/providers/bedrock.py:201]: Similar to https://github.com/mlflow/mlflow/pull/9598/files#r1326549811, could we factor out response translation logic from anthropic, etc. (e.g. https://github.com/mlflow/mlflow/blob/61d8622dde16578dd844ecc12668580dd81a311c/mlflow/gateway/providers/anthropic.py#L84) into a separate method and call it here?

File: docs/source/gateway/index.rst [hunk 3]
```diff
@@ -240,19 +240,19 @@ With the rapid development of LLMs, there is no guarantee that this list will be
...
-+--------------------+--------------------------+--------------------+------------------+-----------------------------+--------------------------+--------------------------+
-| llm/v1/embeddings  | - text-embedding-ada-002 | - instructor-large | N/A              | - embed-english-v2.0        | - text-embedding-ada-002 | - MLflow served models** |
-|                    |                          | - instructor-xl    |                  | - embed-multilingual-v2.0   |                          |                          |
-+--------------------+--------------------------+--------------------+------------------+-----------------------------+--------------------------+--------------------------+
++--------------------+--------------------------+--------------------+------------------+-----------------------------+--------------------------+--------------------------+--------------------------+
+| Route Type         | OpenAI                   | MosaicML           | Anthropic        | Cohere                      | Azure OpenAI             | MLflow                   | AWS Bedrock              |
```
Comment (dbczumar) [inline docs/source/gateway/index.rst:244]: Thanks so much for adding this to docs! Can we add an example in https://github.com/mlflow/mlflow/tree/master/examples/gateway as well? This is awesome!

File: mlflow/gateway/providers/bedrock.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            "AWS_REGION", os.environ.get("AWS_DEFAULT_REGION")
+        )
+
+        if region_name:
+            config_args["region_name"] = session_args["region_name"] = region_name
+
+        # check profile, mostly for dev
+        profile_name = os.environ.get("AWS_PROFILE")
+        if profile_name:
+            session_args["profile_name"] = profile_name
```
Comment (dbczumar) [inline mlflow/gateway/providers/bedrock.py:74]: Seems like this will override what's currently in the boto config. Should we add "profile" to the config YAML instead?

**Comment by andrew-christianson** [comment]
@dbczumar I'll factor out the pieces common to the underlying model.  I think that'd mean a bit of new abstraction (i.e. the 1P and 3P (Bedrock) Anthropic providers calling a common 'AnthropicModelInterface' or some such.

File: mlflow/gateway/providers/bedrock.py (hunk 4 above)
Comment (andrew-christianson) [reply mlflow/gateway/providers/bedrock.py:74]: @dbczumar I'm not sure I follow.  You're thinking this could/should live in the route definition, and be picked up by AWSBedrockConfig like the key/role variables?  Or is there another point at which Gateway is setting / picking up an AWS profile that I'm unknowingly clobbering here?

File: docs/source/gateway/index.rst [hunk 5]
```diff
@@ -537,6 +539,51 @@ Anthropic
 | **anthropic_api_key**   | Yes      | N/A                      | This is the API key for the Anthropic service.        |
 +-------------------------+----------+--------------------------+-------------------------------------------------------+
 
+AWS Bedrock
++++++++++++
+
+Top-level model configuration for AWS Bedrock routes must be one of the two supported two authentication modes: key-based and role-based.
```
Comment (BenWilson2) [inline docs/source/gateway/index.rst:545]:
```suggestion
Top-level model configuration for AWS Bedrock routes must be one of the following two supported authentication modes: `key-based` or `role-based`.
```

File: docs/source/gateway/index.rst [hunk 6]
```diff
@@ -537,6 +539,51 @@ Anthropic
...
+
++--------------------------+----------+------------------------------+-------------------------------------------------------+
+| Configuration Parameter  | Required | Default                      | Description                                           |
++==========================+==========+==============================+=======================================================+
+| **aws_config**           | No       |                              | An object with either the key-based or role-based     |
+|                          |          |                              | schema below.                                         |
++--------------------------+----------+------------------------------+-------------------------------------------------------+
+
+
+To use key-based authentication, define an AWS Bedrock route with usual ``aws_*`` settings.  For testing/development purposes these can be long-lived keys to a and IAM User or short-lived keys to an IAM role.  For long-term use, these *must* be long-lived keys to an IAM user authorized to use Bedrock.
```
Comment (BenWilson2) [inline docs/source/gateway/index.rst:555]:
```suggestion
To use key-based authentication, define an AWS Bedrock route with the required fields below. 

.. note::
  
  If using a configured route purely for development or testing, utilizing an IAM User role or a temporary short-lived standard IAM role are recommended; while for production deployments, a standard long-expiry IAM role is recommended to ensure that the route is capable of handling authentication for a long period. If the authentication expires and a new set of keys need to be supplied, the route must be recreated in order to persist the new keys.
```

File: docs/source/gateway/index.rst [hunk 7]
```diff
@@ -537,6 +539,51 @@ Anthropic
...
+|                          |          |                              | authorized to use bedrock                             |
++--------------------------+----------+------------------------------+-------------------------------------------------------+
+| **aws_access_key_id**    | Yes      |                              | AWS access key ID for the IAM user/role               |
+|                          |          |                              | authorized to use Bedrock                             |
++--------------------------+----------+------------------------------+-------------------------------------------------------+
+| **aws_session_token**    | No       | None                         | Optional session token, if required                   |
++--------------------------+----------+------------------------------+-------------------------------------------------------+
+
+Alternatively, for role-based authentication, define AWS Bedrock route with an a IAM Role authorized to access Bedrock.  AI Gateway will attempt to assume this role with using the standard credential provider chain, and will renew the role credentials if they're expired.
```
Comment (BenWilson2) [inline docs/source/gateway/index.rst:572]:
```suggestion
Alternatively, for role-based authentication, an AWS Bedrock route can be defined and initialized with an a IAM Role  ARN that is authorized to access Bedrock.  The MLflow AI Gateway will attempt to assume this role with using the standard credential provider chain and will renew the role credentials if they have expired.
```

File: mlflow/gateway/config.py [hunk 8]
```diff
@@ -160,11 +161,32 @@ class MlflowModelServingConfig(ConfigModel):
     model_server_url: str
 
 
+class AWSBaseConfig(pydantic.BaseModel):
+    aws_region: Optional[str] = None
+
+
+class AWSRole(AWSBaseConfig):
+    aws_role_arn: str
+    session_length_seconds: int = 15 * 60
```
Comment (BenWilson2) [inline mlflow/gateway/config.py:170]: How important is it to provide a default value for session length that is shorter than the default provided by ARN registration through AWS? The default is 1 hour per the AWS docs. 
If we set the default to be the same as their service, that seems ok (if it's not something that people generally override), but since the default here is shorter than that, it seems like this should be admin-controlled during route configuration.

File: mlflow/gateway/providers/bedrock.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+        )
+
+    def get_bedrock_client(self):
+        if self._client is not None and not self._client_expired():
+            return self._client
+
+        session_args, config_args, client_args = {}, {}, {}
+        aws_config = self.bedrock_config.aws_config
+
+        region_name = aws_config.aws_region or os.environ.get(
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:64]: Could these environment variables be migrated to `environment_variables.py` (https://github.com/mlflow/mlflow/blob/master/mlflow/environment_variables.py) and have a distinct naming convention that identifies what they are used for `MLFLOW_AI_GATEWAY_BEDROCK_AWS_REGION` `MLFLOW_AI_GATEWAY_BEDROCK_AWS_DEFAULT_REGION` ?

Comment (BenWilson2) [reply mlflow/gateway/providers/bedrock.py:64]: Actually, on second thought - why are these able to be overridden in the Provider definition? Is this intended? Similar logic for handling env var defined api keys for other providers is handled within the route constructor validation within `config.py` so that the route definitions remain immutable. 
Is there a distinct need to be mutate the route handling within the Gateway Server after a route has been defined in order to change the AWS Region?

File: mlflow/gateway/providers/bedrock.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+            )
+            client_args["aws_access_key_id"] = role["Credentials"]["AccessKeyId"]
+            client_args["aws_secret_access_key"] = role["Credentials"]["SecretAccessKey"]
+            client_args["aws_session_token"] = role["Credentials"]["SessionToken"]
+        elif isinstance(aws_config, AWSIdAndKey):
+            client_args["aws_access_key_id"] = aws_config.aws_access_key_id
+            client_args["aws_secret_access_key"] = aws_config.aws_secret_access_key
+            client_args["aws_session_token"] = aws_config.aws_session_token
+        else:
+            # defer to standard CredentialProviderChain
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:92]: If there are issues with the CredentialProviderChain, what sort of Exception is raised? Will the server pass this issue correctly back to the client with an informative message about why a 5xx error occurred within the AI Gateway?

File: mlflow/gateway/providers/bedrock.py [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    def _client_expired(self):
+        if not isinstance(self.bedrock_config.aws_config, AWSRole):
+            return False
+
+        return (
+            (time.monotonic_ns() - self._client_created)
+            >= (self.bedrock_config.aws_config.session_length_seconds) * 1_000_000_000,
+        )
+
+    def get_bedrock_client(self):
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:57]: This function is a bit hard to read and contains a lot of branching logic that may be hard to troubleshoot and test. Could this be refactored in a manner like this?

```python

def get_bedrock_client(self):
    if self._client is not None and not self._client_expired():
        return self._client

    session_args = self._construct_session_args()
    session = boto3.Session(**session_args)

    client_args = self._construct_client_args(session, self.bedrock_config)

    try:
        self._client, self._client_created = (
            session.client(service_name="bedrock", **client_args),
            time.monotonic_ns(),
        )
        return self._client
    except botocore.exceptions.UnknownServiceError as e:
        raise AIGatewayConfigException(
            "Cannot create AWS Bedrock client; ensure boto3/botocore "
            "linked from the AWS Bedrock user guide are installed. "
            "Otherwise likely missing credentials or accessing account without to "
            "AWS Bedrock Private Preview"
        ) from e

# Allow for isolating the session arguments
def _construct_session_args(self):
    aws_config = self.bedrock_config.aws_config
    region_name = aws_config.aws_region or os.getenv("AWS_REGION", os.getenv("AWS_DEFAULT_REGION"))
    profile_name = os.getenv("AWS_PROFILE")

    return {k: v for k, v in {
        "region_name": region_name,
        "profile_name": profile_name
    }.items() if v}

# simplify the conditional logic and use dict.update() instead of direct assignment
def _construct_client_args(self, session, aws_config):
    client_args = {}

    if isinstance(aws_config, AWSRole):
        role = session.client(service_name="sts").assume_role(
            RoleArn=aws_config.aws_role_arn,
            RoleSessionName="ai-gateway-bedrock",
            DurationSeconds=aws_config.session_length_seconds,
        )
        client_args.update({
            "aws_access_key_id": role["Credentials"]["AccessKeyId"],
            "aws_secret_access_key": role["Credentials"]["SecretAccessKey"],
            "aws_session_token": role["Credentials"]["SessionToken"]
        })
    elif isinstance(aws_config, AWSIdAndKey):
        client_args.update({
            "aws_access_key_id": aws_config.aws_access_key_id,
            "aws_secret_access_key": aws_config.aws_secret_access_key,
            "aws_session_token": aws_config.aws_session_token
        })
    # remove the else block (unreachable code for the pass, which does nothing)

    return client_args
```

File: mlflow/gateway/providers/bedrock.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+                "linked from the AWS Bedrock user guide are installed. "
+                "Otherwise likely missing credentials or accessing account without to "
+                "AWS Bedrock Private Preview"
+            ) from e
+
+    @property
+    def _underlying_provider(self):
+        if (not self.config.model.name) or "." not in self.config.model.name:
+            return None
+        provider, *_ = self.config.model.name.split(".")
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:118]: If we're just interested in the first element of the split, can we simplify this without providing an anonymous list capture here?

```python
@property
def _underlying_provider(self):
    model_name = self.config.model.name
    if not model_name or "." not in model_name:
        return None

    provider = model_name.split(".")[0]
    return AWSBedrockModelProvider.of_str(provider)
```

File: mlflow/gateway/providers/bedrock.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        return AWSBedrockModelProvider.of_str(provider)
+
+    def _prepare_input(self, payload: dict):
+        prompt, preped = payload.pop("prompt"), {}
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:122]: Can we avoid multiple assignment for unrelated declarations? 
While this is supported in Python, it decreases code readability and makes statements a bit more hard to follow for other contributors who may not use this shorthand style. 

```python
prompt = payload.pop("prompt")
preped = {}
```

File: mlflow/gateway/providers/bedrock.py [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+        # TODO handle top_p, top_k, etc.
+
+        if self._underlying_provider is AWSBedrockModelProvider.AMAZON:
+            preped["inputText"] = prompt
+            preped["textGenerationConfig"] = rename_payload_keys(
+                payload, {"max_tokens": "maxTokenCount", "stop": "stopSequences"}
+            )
+        elif self._underlying_provider is AWSBedrockModelProvider.ANTHROPIC:
+            if not prompt.startswith("Human: "):
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:132]: Can we break out the logic here and in the Anthropic provider to a utility function that both can share? (provided that the logic is identical and Bedrock is merely passing requests through; I'm sorry that I can't verify this as I don't have access to their private API documentation)

File: mlflow/gateway/providers/bedrock.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+                detail=f"Unknown AWS Bedrock model type {self._underlying_provider}",
+            )
+
+    def _make_request(self, body):
+        try:
+            response = self.get_bedrock_client().invoke_model(
+                body=json.dumps(body).encode(),
+                modelId=self.config.model.name,
+                # defaults
+                # save=False,
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:243]: Is this being worked on / intending to be supported? 
Is the mechanism, if set to `True`, to store the request and response to s3? What additional information would need to be provided during route creation to support this?

File: mlflow/gateway/providers/bedrock.py [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+                accept="application/json",
+                contentType="application/json",
+            )
+            return json.loads(response.get("body").read())
+        # except botocore.exceptions.ValidationException as e:
```
Comment (BenWilson2) [inline mlflow/gateway/providers/bedrock.py:248]: Remove, if not needed

File: tests/gateway/providers/test_bedrock.py [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+                        {
+                            "generatedToken": {
+                                "token": "▁It▁looks▁like",
+                                "logprob": -3.2340049743652344,
+                                "raw_logprob": -3.2340049743652344,
+                            },
+                            "topTokens": None,
+                            "textRange": {"start": 1, "end": 14},
+                        },
+                        {
```
Comment (BenWilson2) [inline tests/gateway/providers/test_bedrock.py:80]: We can probably truncate this to just two tokens :)

File: tests/gateway/providers/test_bedrock.py [hunk 18]
```diff
@@ -0,0 +1,480 @@
...
+
+bedrock_live_aws_configs = [
+    {"aws_region": "us-east-1"},
+    pytest.param(
+        {
+            "aws_access_key_id": os.environ.get("AWS_ACCESS_KEY_ID"),
+            "aws_secret_access_key": os.environ.get("AWS_SECRET_ACCESS_KEY"),
+            "aws_session_token": os.environ.get("AWS_SESSION_TOKEN"),
+        },
+        marks=pytest.mark.skipif(
```
Comment (BenWilson2) [inline tests/gateway/providers/test_bedrock.py:324]: Why do we need to skip these tests? 
Since all tests apart from the last test (which is an integration test and shouldn't be in this suite) are mocking portions of access, we should be using `monkeypatch.setenv()` for the `AWS_x` environment keys so that the patched calls can be evaluated.

File: tests/gateway/providers/test_bedrock.py [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+            fix["request"],
+            fix["expected"],
+            marks=[]
+            if fix["provider"] is not AWSBedrockModelProvider.COHERE
+            else pytest.mark.skip("Cohere isn't availabe on AWS Bedrock yet"),
+        )
+        for fix in bedrock_model_provider_fixtures
+    ],
+)
+async def test_live_call_to_aws(provider, config, payload, expected, aws_config):
```
Comment (BenWilson2) [inline tests/gateway/providers/test_bedrock.py:472]: This test should be removed to `test_integration.py` and the outside call to AWS should be mocked. We're not trying to test that AWS can return the results from the language model services, but rather that the client and fluent APIs are capable of communicating up to the point of calling and receiving a response from a provider (in this case, AWS). 
Please have a look at the server constructor fixtures referenced within `test_integration` and see how we have mocked the external calls to other providers for guidance.

File: mlflow/gateway/config.py (hunk 8 above)
Comment (andrew-christianson) [reply mlflow/gateway/config.py:170]: @BenWilson2 Not terribly.  It's weakly important that it be configured to be bit shorter than whatever the AWS-configured maximum duration is, so that a request arriving right at expiry doesn't race `AWSBedrockProvider._client_expired` against the actual expiration.  A much shorter default than theirs is a strong guarantee of that condition :P.  

I can add some docs to that effect.

File: mlflow/gateway/providers/bedrock.py (hunk 11 above)
Comment (andrew-christianson) [reply mlflow/gateway/providers/bedrock.py:57]: Sure, that looks good.  Not sure that dict.update adds a lot there, but it will look nice when you guys drop 3.7 and convert all those to `|=` :p

File: mlflow/gateway/providers/bedrock.py (hunk 15 above)
Comment (andrew-christianson) [reply mlflow/gateway/providers/bedrock.py:243]: No intention of supporting it at this time.

File: mlflow/gateway/providers/bedrock.py (hunk 16 above)
Comment (andrew-christianson) [reply mlflow/gateway/providers/bedrock.py:248]: I would like to catch it to raise as a 422, but botocore exceptions are a PITA to catch.  I'll add a TODO

File: mlflow/gateway/providers/bedrock.py (hunk 9 above)
Comment (andrew-christianson) [reply mlflow/gateway/providers/bedrock.py:64]: The goal here is for defer as much as possible to the normal boto credential provider chain for the benefit of the AWSRole path, while still allowing a RouteConfig to provide credentials.  Given that, I think we can just remove all the explicit AWS_* environment variable stuff and count on boto to Do The Right Thing.

**Comment by andrew-christianson** [comment]
@dbczumar @BenWilson2 all comments should be addressed.  Also Bedrock is GA as of [today](https://www.aboutamazon.com/news/aws/aws-amazon-bedrock-general-availability-generative-ai-innovations)!

File: mlflow/gateway/config.py (hunk 8 above)
Comment (BenWilson2) [reply mlflow/gateway/config.py:170]: Adding this info in docs seems good to me!

File: mlflow/gateway/providers/bedrock.py (hunk 9 above)
Comment (BenWilson2) [reply mlflow/gateway/providers/bedrock.py]: Sounds great!

File: mlflow/gateway/config.py (hunk 8 above)
Comment (BenWilson2) [reply mlflow/gateway/config.py:170]: (Don't worry about this for this PR - I'll file a follow-up PR to add this to the docs)

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for all the hard work here @andrew-christianson it looks great! :D

File: requirements/gateway-requirements.yaml [hunk 20]
```diff
@@ -25,3 +25,8 @@ watchfiles:
 aiohttp:
   pip_release: aiohttp
   max_major_version: 3
+
+boto3:
+  pip_release: boto3
+  minimum: "1.28.56"
+  max_major_version: 2
```
Comment (BenWilson2) [inline requirements/gateway-requirements.yaml:32]:
```suggestion
  max_major_version: 1
```

Sorry that this is a little confusing. It's an inclusive check. 
`max_major_version: 2` translates to `<3`

File: requirements/gateway-requirements.txt [hunk 21]
```diff
@@ -7,3 +7,4 @@ fastapi<1
 uvicorn[standard]<1
 watchfiles<1
 aiohttp<4
+boto3<2,>=1.28
```
Comment (BenWilson2) [inline requirements/gateway-requirements.txt:10]:
```suggestion
boto3<2,>=1.28.56
```

**Comment by BenWilson2** [comment]
Don't worry about the cross version tests failing here. Simply an artifact of the base master branch that your fork PR comes from.

**Comment by BenWilson2** [comment]
Line 211 in gateway integration tests just needs to have its count incremented, since you added a new bedrock route :)

**Comment by BenWilson2** [comment]
To fix the linting issues, the pre-commit hook is super helpful: 

(install from root of forked repo)
``` bash
pre-commit install -t pre-commit -t prepare-commit-msg
```

To run:
``` bash
pre-commit run --all-files --color=always
```

Ruff will fix most of these things that the linter is complaining about :)

File: docs/source/gateway/index.rst
Comment (harupy) [inline docs/source/gateway/index.rst:1]: can we add a minimum example in `examples/gateway`?

File: mlflow/gateway/providers/base.py [hunk 22]
```diff
@@ -35,3 +35,31 @@ def check_for_model_field(payload):
...
+    @abstractclassmethod
+    def model_to_completions(cls, resp, config):
+        raise NotImplementedError
+
+    @abstractclassmethod
+    def completions_to_model(cls, payload, config):
+        raise NotImplementedError
+
+    @abstractclassmethod
+    def embeddings_to_model(cls, payload, config):
```
Comment (harupy) [inline mlflow/gateway/providers/base.py:56]: can we add a docstring to describe what needs to be implemented?

**Comment by BenWilson2** [comment]
@andrew-christianson just a few merge conflicts due to some additions to Gateway since the last time that you rebased. Once those are buttoned up, we'll be good to go to merge and get this included in 2.8! :)

**Review by dbczumar — APPROVED** [review]
